docs/environment.md
# Environment Specifications

This section provides additional information regarding the environment implemented in ALE.

## Available Actions

The following regular actions are defined by the `Action` enum in [`common/Constants.h`](https://github.com/mgbellemare/Arcade-Learning-Environment/blob/master/src/common/Constants.h). These can also be accessed in Python through the enum object `ale_py.Action`. These actions are interpreted by ALE as follows:

| Index | Action | Description |
| ---   | ---    | ---         |
| 0     | **`NOOP`**   | No operation, do nothing. |
| 1     | **`FIRE`**   | Press the fire button without updating the joystick position |
| 2     | **`UP`**    | Apply a Δ-movement upwards on the joystick |
| 3     | **`RIGHT`**    | Apply a Δ-movement rightward on the joystick |
| 4     | **`LEFT`** | Apply a Δ-movement leftward on the joystick |
| 5     | **`DOWN`** | Apply a Δ-movement downward on the joystick |
| 6     | **`UPRIGHT`** | Execute **`UP`** and **`RIGHT`** |
| 7     | **`UPLEFT`** | Execute **`UP`** and **`LEFT`** |
| 8     | **`DOWNRIGHT`** | Execute **`DOWN`** and **`RIGHT`** |
| 9     | **`DOWNLEFT`** | Execute **`DOWN`** and **`LEFT`** |
| 10     | **`UPFIRE`** | Execute **`UP`** and **`FIRE`** |
| 11     | **`RIGHTFIRE`** | Execute **`RIGHT`** and **`FIRE`** |
| 12     | **`LEFTFIRE`** | Execute **`LEFT`** and **`FIRE`** |
| 13     | **`DOWNFIRE`** | Execute **`DOWN`** and **`FIRE`** |
| 14     | **`UPRIGHTFIRE`** | Execute **`UP`** and **`RIGHT`** and **`FIRE`** |
| 15     | **`UPLEFTFIRE`** | Execute **`UP`** and **`LEFT`** and **`FIRE`** |
| 16     | **`DOWNRIGHTFIRE`** | Execute **`DOWN`** and **`RIGHT`** and **`FIRE`** |
| 17     | **`DOWNLEFTFIRE`** | Execute **`DOWN`** and **`LEFT`** and **`FIRE`** |
| 40     | **`RESET`**<sup>1</sup> | Toggles the Atari 2600 reset switch, **not** used for resetting the environment |

<small>1</small>: Note that the **`RESET`** action toggles the Atari 2600 reset switch, rather than reset the
environment, and as such is ignored by most interfaces.

Note: There are two main types of controllers on the Atari 2600 console. The [joystick controller](https://en.wikipedia.org/wiki/Atari_CX40_joystick) and the [paddle controller](https://en.wikipedia.org/wiki/Paddle_\(game_controller\)). For paddle controllers all **`*RIGHT*`** actions correspond to a Δ-movment to the right on the wheel, and all **`*LEFT*`** actions correspond to a Δ-movement to the left.


##  Terminal States

Once the end of episode is reached (a terminal state in RL terminology), no further emulation
takes place until the appropriate reset command is sent. This command is distinct from the Atari
2600 reset. This "system reset" avoids odd situations where the player can reset the game
through button presses, or where the game normally resets itself after a number of frames. This
makes for a cleaner environment interface. The interfaces described here all provide a system reset command or method.

## Color Averaging

Many Atari 2600 games display objects on alternating frames (sometimes even less frequently).
This can be an issue for agents that do not consider the whole screen history.
By default, _color averaging_ is **not** enabled, that is, the environment output is the actual frame from the emulator.
This behaviour can be turned on using `setBool` with the `color_averaging` key.

## Action Repeat Stochasticity

Beginning with ALE 0.5.0, there is now an option (enabled by default) to add
_action repeat stochasticity_ to the environment. With probability 𝗉 (default: 𝗉 = 0.25),
the previously executed action is executed again during the next frame, ignoring the agent's
actual choice. This value can be modified using the option `action_repeat_probability`.
The default value was chosen as the highest value for which human play-testers
were unable to detect any delay or control lag. ([Machado et al. 2018](#references-machado18)).

The motivation for introducing action repeat stochasticity was to help separate _trajectory optimization_ research from _robust controller optimization_, the latter often being the
desired outcome in reinforcement learning (RL). We strongly encourage RL researchers to use
the default stochasticity level in their agents, and clearly report the setting used.

## Minimal Action Set

It may sometimes be convenient to restrict the agent to a smaller action set. This can be
accomplished by querying the `RomSettings` class using the method
`getMinimalActionSet`. This then returns a set of actions judged "minimal" to play a given
game. Due to the potentially high impact of this setting on performance, we encourage researchers
to clearly report the method used in their experiments.


## Modes and Difficulties

ALE 0.6.0 introduces modes and difficulties, which can be set using the relevant methods `setMode`, `setDifficulty`. These introduce a whole range of new environments. For more details, see [Machado et al. 2018](#references-machado18).


## References

<sup><a name="references-machado18">1</a></sup> <small>**Revisiting the Arcade Learning Environment: Evaluation Protocols and Open Problems for General Agents**</small>
<small>M. C. Machado, M. G. Bellemare, E. Talvitie, J. Veness, M. J. Hausknecht, M. Bowling., Journal of Artificial Intelligence Research, Volume 61, pages 523-562, 2018. URL: [https://jair.org/index.php/jair/article/view/11182](https://jair.org/index.php/jair/article/view/11182)</small>


ChangeLog
Inter-release notes:
  * color_averaging is now off by default so that environment observations correspond to emulator frames unless requested otherwise.

October 4th, 2015. ALE 0.5dev_b.
  * Enforce flags existence (@mcmachado).
  * Fix RNG issues introduced in 0.5.0.
  * Routines for ALEState serialization (@Jragonmiris). 
  * Additional bug fixes.

July 7th, 2015. ALE 0.5dev.
  * Refactored Python getScreenRGB to return unpacked RGB values (@spragunr).
  * Sets the default value of the color_averaging flag to be true. It was true by default in previous versions but was changed in 0.5.0. Reverted for backward compatibility.
  * Added RNG serialization capability.
  * Bug fixes from ALE 0.5.0.

June 22nd, 2015. ALE 0.5.0.
  * Added action_repeat_stochasticity.
  * Added sound playback, visualization.
  * Added screen/sound recording ability.
  * CMake now available.
  * Incorporated Benjamin Goodrich's Python interface.
  * Some game fixes.
  * Added examples for shared library, Python, fifo, RL-Glue interfaces.
  * Incorporated Java agent into main repository.
  * Removed internal controller, now superseded by shared library interface.
  * Better ALEInterface.
  * Many other changes.

February 15th, 2015. ALE 0.5dev.
  * Removed the following command-line flags: 'output_file', 'system_reset_steps', 'use_environment_distribution', 'backward_compatible_save', internal agent flags
  * The flag 'use_starting_actions' was removed and internally its value is always 'true'.
  * The flag 'disable_color_averaging' was renamed to 'color_averaging' and FALSE is its default value.

April 28th, 2014. ALE 0.4.4.
  * Fixed a memory issue in ALEScreen.

April 26th, 2014. Bug fix (Mayank Daswani).
  * Fixed issues with frame numbers not being correctly updated.
  * Fixed a bug where total reward was not properly reported under frame skipping. 

January 7th, 2013. ALE 0.4.3.
  * Fixed a bug with ALEState's m_frame_number.

June 12th, 2013. ALE 0.4.2. 
  * Modified StellaEnvironment save/load interface to provide additional flexibility. 
  * Series of bug fixes from Matthew Hausknecht and community.

May 24th, 2013. Bug fix to ALE 0.4.1.
  * Fixed RL-Glue syntax from OBSERVATION to OBSERVATIONS. Thanks to Angus MacIsaac for picking this bug up.

May 22nd, 2013. ALE 0.4.1.
  * Added frame skipping support directly in StellaEnvironment.
  * Reverted default number of episodes to 10.

April 22nd, 2013. ALE 0.4.0.
  * RL-Glue support 
  * Shared library interface
  * Simpler direct environment interfacing
  * Improved environment handling
  * Improved environment customization 
  * Better documentation

October 3rd, 2012. Moving to GITHub.

August 7th, 2012. ALE 0.3.1.
  * Fixed frames per episode cap for FIFO agents, added functionality to
    limit the total number of frames per run for FIFO agents.

July 22nd, 2012. ALE 0.3.
  * Initial ALE release.



src/emucore/Settings.cxx
//============================================================================
//
//   SSSS    tt          lll  lll
//  SS  SS   tt           ll   ll
//  SS     tttttt  eeee   ll   ll   aaaa
//   SSSS    tt   ee  ee  ll   ll      aa
//      SS   tt   eeeeee  ll   ll   aaaaa  --  "An Atari 2600 VCS Emulator"
//  SS  SS   tt   ee      ll   ll  aa  aa
//   SSSS     ttt  eeeee llll llll  aaaaa
//
// Copyright (c) 1995-2007 by Bradford W. Mott and the Stella team
//
// See the file "license" for information on usage and redistribution of
// this file, and for a DISCLAIMER OF ALL WARRANTIES.
//
// $Id: Settings.cxx,v 1.125 2007/08/22 13:55:40 stephena Exp $
//============================================================================

#include <cassert>
#include <sstream>
#include <fstream>
#include <algorithm>
#include <string>

#include "emucore/OSystem.hxx"
#include "emucore/Settings.hxx"

namespace ale {
namespace stella {

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Settings::Settings(OSystem* osystem) : myOSystem(osystem) {
    // Add this settings object to the OSystem
    myOSystem->attach(this);

    setInternal("palette", "standard");

    setInternal("sound", "false");
    setInternal("fragsize", "512");
    setInternal("freq", "31400");
    setInternal("tiafreq", "31400");
    setInternal("volume", "100");
    setInternal("clipvol", "true");

    setDefaultSettings();
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Settings::~Settings()
{
  myInternalSettings.clear();
  myExternalSettings.clear();
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::validate()
{
  std::string s;
  int i;

#ifdef SDL_SUPPORT
  i = getInt("volume");
  if(i < 0 || i > 100)
    setInternal("volume", "100");
  i = getInt("freq");
  if(i < 0 || i > 48000)
    setInternal("freq", "31400");
  i = getInt("tiafreq");
  if(i < 0 || i > 48000)
    setInternal("tiafreq", "31400");
#endif

  s = getString("palette");
  if(s != "standard" && s != "z26" && s != "user")
    setInternal("palette", "standard");
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::setInt(const std::string& key, const int value)
{
  std::ostringstream stream;
  stream << value;

  if(int idx = getInternalPos(key) != -1){
    setInternal(key, stream.str(), idx);
  }
  else{
    verifyVariableExistence(intSettings, key);
    setExternal(key, stream.str());
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::setFloat(const std::string& key, const float value)
{
  std::ostringstream stream;
  stream << value;

  if(int idx = getInternalPos(key) != -1){
    setInternal(key, stream.str(), idx);
  }
  else{
    verifyVariableExistence(floatSettings, key);
    setExternal(key, stream.str());
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::setBool(const std::string& key, const bool value)
{
  std::ostringstream stream;
  stream << value;

  if(int idx = getInternalPos(key) != -1){
    setInternal(key, stream.str(), idx);
  }
  else{
    verifyVariableExistence(boolSettings, key);
    setExternal(key, stream.str());
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::setString(const std::string& key, const std::string& value)
{
  if(int idx = getInternalPos(key) != -1){
    setInternal(key, value, idx);
  }
  else{
    verifyVariableExistence(stringSettings, key);
    setExternal(key, value);
  }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::getSize(const std::string& key, int& x, int& y) const
{
  std::string size = getString(key);
  replace(size.begin(), size.end(), 'x', ' ');
  std::istringstream buf(size);
  buf >> x;
  buf >> y;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int Settings::getInt(const std::string& key, bool strict) const {
    // Try to find the named setting and answer its value
    int idx = -1;
    if((idx = getInternalPos(key)) != -1) {
        return (int) atoi(myInternalSettings[idx].value.c_str());
    } else {
        if((idx = getExternalPos(key)) != -1) {
            return (int) atoi(myExternalSettings[idx].value.c_str());
        } else {
            if (strict) {
                ale::Logger::Error << "No value found for key: " << key << ". ";
                ale::Logger::Error << "Make sure all the settings files are loaded." << std::endl;
                exit(-1);
            } else {
                return -1;
            }
        }
    }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
float Settings::getFloat(const std::string& key, bool strict) const {
    // Try to find the named setting and answer its value
    int idx = -1;
    if((idx = getInternalPos(key)) != -1) {
        return (float) atof(myInternalSettings[idx].value.c_str());
    } else {
        if((idx = getExternalPos(key)) != -1) {
            return (float) atof(myExternalSettings[idx].value.c_str());
        } else {
            if (strict) {
                ale::Logger::Error << "No value found for key: " << key << ". ";
                ale::Logger::Error << "Make sure all the settings files are loaded." << std::endl;
                exit(-1);
            } else {
                return -1.0;
            }
        }
    }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
bool Settings::getBool(const std::string& key, bool strict) const {
    // Try to find the named setting and answer its value
    int idx = -1;
    if((idx = getInternalPos(key)) != -1)
    {
        const std::string& value = myInternalSettings[idx].value;
        if(value == "1" || value == "true" || value == "True")
            return true;
        else if(value == "0" || value == "false" || value == "False")
            return false;
        else
            return false;
    } else if((idx = getExternalPos(key)) != -1) {
        const std::string& value = myExternalSettings[idx].value;
        if(value == "1" || value == "true")
            return true;
        else if(value == "0" || value == "false")
            return false;
        else
            return false;
    } else {
        if (strict) {
            ale::Logger::Error << "No value found for key: " << key << ". ";
            ale::Logger::Error << "Make sure all the settings files are loaded." << std::endl;
            exit(-1);
        } else {
            return false;
        }
    }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
const std::string& Settings::getString(const std::string& key, bool strict) const {
    // Try to find the named setting and answer its value
    int idx = -1;
    if((idx = getInternalPos(key)) != -1) {
        return myInternalSettings[idx].value;
    } else if ((idx = getExternalPos(key)) != -1) {
        return myExternalSettings[idx].value;
    } else {
        if (strict) {
            ale::Logger::Error << "No value found for key: " << key << ". ";
            ale::Logger::Error << "Make sure all the settings files are loaded." << std::endl;
            exit(-1);
        } else {
            static std::string EmptyString("");
            return EmptyString;
        }
    }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::setSize(const std::string& key, const int value1, const int value2)
{
  std::ostringstream buf;
  buf << value1 << "x" << value2;
  setString(key, buf.str());
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int Settings::getInternalPos(const std::string& key) const
{
  for(unsigned int i = 0; i < myInternalSettings.size(); ++i)
    if(myInternalSettings[i].key == key)
      return i;

  return -1;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int Settings::getExternalPos(const std::string& key) const
{
  for(unsigned int i = 0; i < myExternalSettings.size(); ++i)
    if(myExternalSettings[i].key == key)
      return i;

  return -1;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int Settings::setInternal(const std::string& key, const std::string& value,
                          int pos, bool useAsInitial)
{
  int idx = -1;

  if(pos != -1 && pos >= 0 && pos < (int)myInternalSettings.size() &&
     myInternalSettings[pos].key == key)
  {
    idx = pos;
  }
  else
  {
    for(unsigned int i = 0; i < myInternalSettings.size(); ++i)
    {
      if(myInternalSettings[i].key == key)
      {
        idx = i;
        break;
      }
    }
  }

  if(idx != -1)
  {
    myInternalSettings[idx].key   = key;
    myInternalSettings[idx].value = value;
    if(useAsInitial) myInternalSettings[idx].initialValue = value;

    /*cerr << "modify internal: key = " << key
         << ", value  = " << value
         << ", ivalue = " << myInternalSettings[idx].initialValue
         << " @ index = " << idx
         << endl;*/
  }
  else
  {
    Setting setting;
    setting.key   = key;
    setting.value = value;
    if(useAsInitial) setting.initialValue = value;

    myInternalSettings.push_back(setting);
    idx = myInternalSettings.size() - 1;

    /*cerr << "insert internal: key = " << key
         << ", value  = " << value
         << ", ivalue = " << setting.initialValue
         << " @ index = " << idx
         << endl;*/
  }

  return idx;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
int Settings::setExternal(const std::string& key, const std::string& value,
                          int pos, bool useAsInitial)
{
  int idx = -1;

  if(pos != -1 && pos >= 0 && pos < (int)myExternalSettings.size() &&
     myExternalSettings[pos].key == key)
  {
    idx = pos;
  }
  else
  {
    for(unsigned int i = 0; i < myExternalSettings.size(); ++i)
    {
      if(myExternalSettings[i].key == key)
      {
        idx = i;
        break;
      }
    }
  }

  if(idx != -1)
  {
    myExternalSettings[idx].key   = key;
    myExternalSettings[idx].value = value;
    if(useAsInitial) myExternalSettings[idx].initialValue = value;

    /*cerr << "modify external: key = " << key
         << ", value = " << value
         << " @ index = " << idx
         << endl;*/
  }
  else
  {
    Setting setting;
    setting.key   = key;
    setting.value = value;
    if(useAsInitial) setting.initialValue = value;

    myExternalSettings.push_back(setting);
    idx = myExternalSettings.size() - 1;

    /*cerr << "insert external: key = " << key
         << ", value = " << value
         << " @ index = " << idx
         << endl;*/
  }

  return idx;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Settings::Settings(const Settings&)
{
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
Settings& Settings::operator = (const Settings&)
{
  assert(false);

  return *this;
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
void Settings::setDefaultSettings() {

    // Stella settings
    stringSettings.insert(std::pair<std::string, std::string>("cpu", "low")); // Reduce CPU emulation fidelity for speed
    // Random seed for ale::stella::System.
    // This random seed should be fixed to enable full determinism in the ALE
    intSettings.insert(std::pair<std::string, int>("system_random_seed", 4753849));

    // Controller settings
    intSettings.insert(std::pair<std::string, int>("max_num_frames", 0));
    intSettings.insert(std::pair<std::string, int>("max_num_frames_per_episode", 0));

    // Expose paddle_min and paddle_max settings but set as 'undefined' so
    // PADDLE_MIN and PADDLE_MAX defines are used as for default values in
    // the StellaEnvironment constructor.
    intSettings.insert(std::pair<std::string, int>("paddle_min", -1));
    intSettings.insert(std::pair<std::string, int>("paddle_max", -1));

    // FIFO controller settings
    boolSettings.insert(std::pair<std::string, bool>("run_length_encoding", true));

    // Environment customization settings
    boolSettings.insert(std::pair<std::string, bool>("restricted_action_set", false));
    intSettings.insert(std::pair<std::string, int>("random_seed", -1));
    boolSettings.insert(std::pair<std::string, bool>("color_averaging", false));
    boolSettings.insert(std::pair<std::string, bool>("send_rgb", false));
    intSettings.insert(std::pair<std::string, int>("frame_skip", 1));
    floatSettings.insert(std::pair<std::string, float>("repeat_action_probability", 0.25));
    stringSettings.insert(std::pair<std::string, std::string>("rom_file", ""));

    // Record settings
    intSettings.insert(std::pair<std::string, int>("fragsize", 64)); // fragsize to 64 ensures proper sound sync
    stringSettings.insert(std::pair<std::string, std::string>("record_screen_dir", ""));
    stringSettings.insert(std::pair<std::string, std::string>("record_sound_filename", ""));

    // Display Settings
    boolSettings.insert(std::pair<std::string, bool>("display_screen", false));

    for(std::map<std::string, std::string>::iterator it = stringSettings.begin(); it != stringSettings.end(); it++) {
      this->setString(it->first, it->second);
    }

    for(std::map<std::string, float>::iterator it = floatSettings.begin(); it != floatSettings.end(); it++) {
      this->setFloat(it->first, it->second);
    }

    for(std::map<std::string, bool>::iterator it = boolSettings.begin(); it != boolSettings.end(); it++) {
      this->setBool(it->first, it->second);
    }

    for(std::map<std::string, int>::iterator it = intSettings.begin(); it != intSettings.end(); it++) {
      this->setInt(it->first, it->second);
    }
}

// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template<typename ValueType>
void Settings::verifyVariableExistence(std::map<std::string, ValueType> dict, std::string key){
    if(dict.find(key) == dict.end()){
      throw std::runtime_error("The key " + key + " you are trying to set does not exist or has incorrect value type.\n");
    }
}

}  // namespace stella
}  // namespace ale
